functions/Start-DbaAgentJob.ps1

```diff
@@ -120,69 +164,72 @@ function Start-DbaAgentJob {
-                            $i = 600
-                        }
+                    # If it failed fast, speed up output
+                    if ($lastrun -ne $currentjob.LastRunDate) {
+                        $i = 600
                     }
+                }
 
-                    # Wait for the job
-                    if (Test-Bound -ParameterName Wait) {
-                        while ($currentjob.CurrentRunStatus -ne 'Idle') {
-                            $currentRunStatus = $currentjob.CurrentRunStatus
-                            $currentStep = $currentjob.CurrentRunStep
-                            $jobStepsCount = $currentjob.JobSteps.Count
-                            $currentStepRetryAttempts = $currentjob.CurrentRunRetryAttempt
-                            if (-not $currentStepRetryAttempts) { $currentStepRetryAttempts = "0" }
-                            $currentStepRetries = $currentjob.RetryAttempts
-                            if (-not $currentStepRetries) { $currentStepRetries = "Unknown" }
-                            Write-Message -Level Verbose -Message "$currentjob is $currentRunStatus, currently on Job Step $currentStep / $jobStepsCount, and has tried $currentStepRetryAttempts / $currentStepRetries retry attempts"
-                            Start-Sleep -Seconds $WaitPeriod
-                            $currentjob.Refresh()
-                        }
-                        Get-DbaAgentJob -SqlInstance $server -Job $currentjob.Name
-                    } else {
-                        Get-DbaAgentJob -SqlInstance $server -Job $currentjob.Name
+                if ((Test-Bound -ParameterName Wait) -and (Test-Bound -not -ParameterName Parallel)) {
+                    # Wait for the job each job in a serialized fashion.
+                    while ($currentjob.CurrentRunStatus -ne 'Idle') {
+                        Invoke-Command -ScriptBlock $waitBlock -ArgumentList @($currentjob, $true, $WaitPeriod)
                     }
+                    Get-DbaAgentJob -SqlInstance $server -Job $($currentjob.Name)
+                } elseif (Test-Bound -not -ParameterName Parallel) {
+                    Get-DbaAgentJob -SqlInstance $server -Job $($currentjob.Name)
+                }
+            }
+        }
+
+        # Wait for each job to be done in parallel
+        if (Test-Bound -ParameterName Parallel) {
+            while ($InputObject.Where( { $PSItem.CurrentRunStatus -ne 'Idle' } ).Count -gt 0) {
+                foreach ($currentjob in $InputObject.Where( { $PSItem.CurrentRunStatus -ne 'Idle' } )) {
+                    Invoke-Command -ScriptBlock $waitBlock -ArgumentList @($currentjob)
                 }
+                Start-Sleep -Seconds $WaitPeriod
             }
+            # We can be sloppy with -Job since Get-DbaAgentJob will filter out per server, just don't query each server more than once.
+            Get-DbaAgentJob -SqlInstance $($InputObject.Parent.Parent | Select-Object -Unique) -Job $InputObject
```

Comment:
Isn't $InputObject an array here? I'm not sure that this would be a valid syntax in multi-server multi-job scenario.

Reply:
I tested it across multiple servers.  Worked like a charm.  But that's why it's also piped to Select-Object -Unique.

Reply:
But I could alway add a ForEach-Object within to guarantee the behavior.

Reply:
I'm surprised it works as Get-DbaAgentJob only accepts -Job as an array of strings. What happens when $InputObject is a piped Job object?
https://github.com/sqlcollaborative/dbatools/blob/development/functions/Get-DbaAgentJob.ps1#L98

If feel like the best solution here would be to teach Get-DbaAgentJob to accept job objects via pipeline. It might, however, be out of scope here, as it would need a few things added:
- add $InputObject parameter with ValueFromPipeline
- make $SqlInstance non-mandatory
- teach it to recognize the $InputObject class similar to:
https://github.com/sqlcollaborative/dbatools/blob/b6f099f5572ff1dbe170f961d4ca2cc031a751ec/functions/Export-DbaRegServer.ps1#L94-L105

Reply:
The other option would be to iterate over $InputObject, that's probably the safest bet for now.